### ISSUE TITLE
Kucoin errorhandling

### ISSUE THREAD

**Issue by gaardiolor** [ccxt/ccxt#1258]
Is this enough info to convert these errors from ExchangeError to NetworkError ? Sorry I don't have verbose logging this time, hopefully the stacktraces give enough info. 
```
exc_type, exc_value, exc_traceback = sys.exc_info()
traceback.print_exception(exc_type, exc_value, exc_traceback, limit=20, file=sys.stdout)
```

- The 502 'Gateway Timeout' seems obvious.. It should be NetworkError by default imo. If exchanges use this error correctly, you can always recover from it by waiting a little (or a lot) and retry. Surely there are some exchanges that use 502 incorrectly, but shouldn't we make exceptions for those then, instead of the other way around ?
- The 404 is less obvious, normally you can't recover from this by retrying. But with Kucoin you can somehow, they unjustly generate 404's when they're busy.

The 502, `fetch_order_book`
```
Traceback (most recent call last):
  File "/usr/local/scripts/test2/test2.py", line 841, in fetch_with_retries
    result = json.loads(json.dumps(await fetch()), use_decimal=True)
  File "/usr/lib/python3.6/site-packages/ccxt/async/kucoin.py", line 285, in fetch_order_book
    }, params))
  File "/usr/lib/python3.6/site-packages/ccxt/async/kucoin.py", line 534, in request
    response = await self.fetch2(path, api, method, params, headers, body)
  File "/usr/lib/python3.6/site-packages/ccxt/async/base/exchange.py", line 88, in fetch2
    return await self.fetch(request['url'], request['method'], request['headers'], request['body'])
  File "/usr/lib/python3.6/site-packages/ccxt/async/base/exchange.py", line 110, in fetch
    self.handle_errors(response.status, text, url, method, None, text)
  File "/usr/lib/python3.6/site-packages/ccxt/async/kucoin.py", line 531, in handle_errors
    raise ExchangeError(self.id + ' ' + str(code) + ' ' + reason)
ccxt.base.errors.ExchangeError: kucoin 502 <html>
<head><title>502 Bad Gateway</title></head>
<body bgcolor="white">
<center><h1>502 Bad Gateway</h1></center>
</body>
</html>
<!-- a padding to disable MSIE and Chrome friendly error page -->
<!-- a padding to disable MSIE and Chrome friendly error page -->
<!-- a padding to disable MSIE and Chrome friendly error page -->
<!-- a padding to disable MSIE and Chrome friendly error page -->
<!-- a padding to disable MSIE and Chrome friendly error page -->
<!-- a padding to disable MSIE and Chrome friendly error page -->
```

The 404, `create_limit_buy_order`
```
Traceback (most recent call last):
  File "/usr/local/scripts/test2/test2.py", line 703, in do_order
    order_output = await action_with_retries(e, lambda: e.create_limit_buy_order(pair, float(amount), float(price)), retries=retries)
  File "/usr/local/scripts/test2/test2.py", line 880, in action_with_retries
    result = await action()
  File "/usr/lib/python3.6/site-packages/ccxt/async/kucoin.py", line 357, in create_order
    response = await self.privatePostOrder(self.extend(order, params))
  File "/usr/lib/python3.6/site-packages/ccxt/async/kucoin.py", line 534, in request
    response = await self.fetch2(path, api, method, params, headers, body)
  File "/usr/lib/python3.6/site-packages/ccxt/async/base/exchange.py", line 88, in fetch2
    return await self.fetch(request['url'], request['method'], request['headers'], request['body'])
  File "/usr/lib/python3.6/site-packages/ccxt/async/base/exchange.py", line 110, in fetch
    self.handle_errors(response.status, text, url, method, None, text)
  File "/usr/lib/python3.6/site-packages/ccxt/async/kucoin.py", line 531, in handle_errors
    raise ExchangeError(self.id + ' ' + str(code) + ' ' + reason)
ccxt.base.errors.ExchangeError: kucoin 404
{}
```

**Comment by kroitor** [comment]
> Is this enough info to convert these errors from ExchangeError to NetworkError ?
> - 502

Yes, this error is usually temporary. In fact it is already handled as a recoverable NetworkError. However, the Kucoin handler redefines it, I'll fix it shortly.

> - 404

No, this is most likely a permanent error (endpoint renamed, API upgraded, etc).

**Comment by gaardiolor** [comment]
Yeah, with most websites it's a permanent error. But with Kucoin it isn't, really.. It just happens randomly at peak hours.

**Comment by kroitor** [comment]
Fixed in 1.10.707. Thx!

**Comment by gaardiolor** [comment]
Cool! Both the 502 and 404 ?

**Comment by kroitor** [comment]
@gaardiolor yes

**Comment by gaardiolor** [comment]
@kroitor unfortunately the 404 is still raised as ExchangeError in v707. Can you please take a look ?

```
Traceback (most recent call last):
  File "/usr/local/scripts/test2/test2.py", line 731, in do_order
    order_output = await action_with_retries(e, lambda: e.create_limit_buy_order(pair, float(amount), float(price)), public=False, retries=retries)
  File "/usr/local/scripts/test2/test2.py", line 868, in action_with_retries
    result = json.loads(json.dumps(await action()), use_decimal=True)
  File "/usr/lib/python3.6/site-packages/ccxt/async/kucoin.py", line 357, in create_order
    response = await self.privatePostOrder(self.extend(order, params))
  File "/usr/lib/python3.6/site-packages/ccxt/async/kucoin.py", line 534, in request
    response = await self.fetch2(path, api, method, params, headers, body)
  File "/usr/lib/python3.6/site-packages/ccxt/async/base/exchange.py", line 88, in fetch2
    return await self.fetch(request['url'], request['method'], request['headers'], request['body'])
  File "/usr/lib/python3.6/site-packages/ccxt/async/base/exchange.py", line 110, in fetch
    self.handle_errors(response.status, text, url, method, None, text)
  File "/usr/lib/python3.6/site-packages/ccxt/async/kucoin.py", line 531, in handle_errors
    raise ExchangeError(self.id + ' ' + str(code) + ' ' + reason)
ccxt.base.errors.ExchangeError: kucoin 404
{}
```

**Comment by kroitor** [comment]
Ok, refixed it once again in 1.10.726. Let us know if you have further difficulties with it.

**Comment by gaardiolor** [comment]
@kroitor I've seen now that sometimes Kucoin returns a 404 on for example `create_limit_sell_order`, but did execute the order in the background. So we cannot safely retry in this case. Not sure what to do here. Maybe leave it like this and in my script I need to fix that for all orders, regardless of the exception, always check if it was executed or not. What do you think ?

**Comment by kroitor** [comment]
> Maybe leave it like this and in my script I need to fix that for all orders, regardless of the exception, always check if it was executed or not. What do you think ?

I would say you're moving in the right direction, in general, you're not guaranteed any result right away from create_order (can be a http error, a net quirk, exchange outage, your local connection problem, a lightning just hit the house, anything...), so, upon an error you have to check the result of execution to make sure. An network error from create_order means literally: *the order may be executed or may be not*.

**Comment by gaardiolor** [comment]
Yes exactly, I agree. I'll fix it in my script. Thanks!

**Comment by Fcl69** [comment]
@gaardiolor Do you mind to share the workaround you end up using for kucoin? 404 is very iterating to say the least...

**Comment by gaardiolor** [comment]
@Fcl69 Sure!
Basically I check for the following exceptions: `(ccxt.RequestTimeout, ccxt.ExchangeError, ccxt.NetworkError, ccxt.InvalidNonce, ccxt.AuthenticationError)` . 

If I get them and I did a read operation, I simply sleep for a second and retry. After 10 tries I abort.

If I get them and I did a write operation (order, cancel_order), It could be that it processed anyway. So I check for that with `fetch_balance()` . If it went through within a minute I don't retry. 

If it appears that it did not went through, it could be lagging and go through later (it made me cry the first time.. but I'm in the acceptance phase now): If it's a trade that I can retry without me getting very sad if I end up doing the same trade twice: I simply retry. Mostly those are trades that would completely empty the source symbol balance. I can retry those as much as I want, the balance can't go below 0 anyway. 
If it appears that it did not went through and it's a trade that would made me sad if it happens twice I'm not retrying it. Periodically I check if it went through or not until I'm pretty sure it didn't (10 minutes or so). 

I've tried quite some exchanges now, and almost all of them are horrible to script against. Error-catching and retrying is the easy part, the tricky part is the fact that the errors themselves are not reliable and that, despite the error, the operation in the background could be successful. After some time. Maybe. It's horrible really.
